### PR TITLE
fix(storage): make MMCE and MX4SIO first entry reliable

### PR DESCRIPTION
**Pull request by NathanNeurotic**
## Report

FifthFox tested the rolling build on **SCPH-70012 / NTSC-U** with MMCE, SD2PS2TD/MX4SIO, and USB:

- MX4SIO stopped at **42%** after one brief device read.
- MMCE failed on the first attempt, then worked on the second.

Both failures occur during device discovery/readiness, before POPStarter launch.

### MX4SIO

The page already retries six times, but each attempt previously discovered the MX4SIO slot by sweeping `mass:/` through `mass9:/`. `doesFolderExist` and `getMassMountDriver` can reach `fileXioDopen`; while `bdmfs_fatfs` owns its mount lock, that native call can block indefinitely. A blocked first attempt never reaches retry two.

### MMCE

`System.ensureMmceman()` reports that the IRX loaded, not that `mmce0:/` or `mmce1:/` is mounted. The existing Lua path immediately treated that as complete readiness and permanently set both `_mmce_ready` and `MMCE.PROBED`, so a transient first miss became a cached failure.

### Authoritative runtime fix (`bin/POPSLDR/system.lua`)

- Add a direct BDM identity builder using `System.bdmList()`, driver name, and `parId`.
- For MX4SIO only, obtain `mass:/` or `massN:/` from that lock-free table instead of opening ten filesystem roots.
- Preserve the existing six-attempt MX4SIO retry flow and add one initial registration settle after successful `initMX4SIO()`.
- Make `EnsureMmceReadyOnce()` distinguish module-load success from filesystem readiness.
- Poll `mmce0:/` and `mmce1:/` through a bounded first-access window.
- Set `_mmce_ready` only after an MMCE root is accessible.
- Reinitialize the SIO2 pad path only once after the successful module load.
- Do not cache an empty MMCE slot result; later page access, launch preflight, hot-plug, and R1 refresh remain retryable.

### MX4SIO boot-path translation (`etc/boot.lua`)

- Replace the boot-time `getMassMountDriver`/`fileXioDopen` sweep with a fresh `System.bdmList()` lookup on each bounded attempt.
- Remove the temporary global wrappers and cached BDM snapshot used by the first draft.

### Regression coverage (`tools/host_harness.py`)

- Add a test proving an MMCE timeout does not latch false readiness and that a later root appearance succeeds without a second pad reinitialization.
- Add a test proving MX4SIO identity uses the BDM table and performs zero bare-root filesystem or native mount-driver probes.

## Review handling

Gemini's stale-cache finding was valid. Rather than retaining the wrapper and narrowing its cache window, the final implementation removes the cached global wrapper architecture entirely. BDM identity is read fresh inside the authoritative MX4SIO paths.

CodeRabbit's docstring-coverage warning was not implemented: it was generated against a temporary Python applicator that was removed from the final diff and was unrelated to the Lua runtime fix.

No review suggestion was implemented without validating it against the actual call flow.

## Preservation and scope

- The eager USB/ATA startup order that fixed internal exFAT HDD remains unchanged.
- USB and ATA identity behavior remains unchanged; the lock-free identity replacement is scoped to MX4SIO.
- POPStarter resolution, BDMA staging, selectors, launch handoff, POPS modules, APA/PFS, SMB, and game-list parsing are untouched.
- Missing MMCE or MX4SIO hardware still returns through bounded failure paths.

## Validation

Dedicated validation run **29746248615** passed all stages against the exact final patch before committing it:

- Host execution and regression battery: **22/22 passed**.
- New MMCE retry/latch regression test: passed.
- New MX4SIO no-filesystem-sweep regression test: passed.
- Lua 5.4 syntax validation: passed.
- Full `make clean elfloader all` PS2 build: passed.
- `git diff --check`: passed.

Real-hardware confirmation remains required for PS2 timing behavior.

## Hardware test matrix

1. Cold boot and enter MMCE once; it should not require a second attempt.
2. Repeat MMCE first access from USB, MC, HDD, MX4SIO, and MMCE boot sources.
3. Cold boot and enter MX4SIO once; it must list games or return a bounded not-detected result, never remain at 42%.
4. Test MX4SIO with USB connected simultaneously and confirm USB is not selected as the MX4SIO root.
5. Test an MX4SIO adapter with no readable card and with no adapter present.
6. Run ten complete cold boots and one R1 rescan after both success and failure.
7. Launch one known title from MMCE, MX4SIO, and USB.
8. Reconfirm internal exFAT HDD and APA/PFS HDD page access and launch.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved MMCE detection so temporary unavailability can be retried successfully.
  * Prevented incorrect readiness states when MMCE storage is not yet mounted.
  * Improved MX4SIO storage detection and reduced startup hangs or incorrect root selection.
  * Added safer handling when the initial boot argument is missing.

* **Tests**
  * Added coverage for MMCE retry behavior and MX4SIO device-based storage discovery.

### CODE REVIEW

**Review by NathanNeurotic — COMMENTED** [review]
The current Gemini finding has been validated and addressed in the branch. Requesting a fresh automated pass now that the PR is ready for review.

**Review by NathanNeurotic — COMMENTED** [review]
@coderabbitai review

**Review by NathanNeurotic — COMMENTED** [review]
@coderabbitai review

**Review by NathanNeurotic — COMMENTED** [review]
@coderabbitai review

**Comment by NathanNeurotic** [comment]
@coderabbitai review

File: etc/boot.lua [hunk 1]
```diff
@@ -34,6 +34,154 @@ end
...
+        return driver
+      end
+      -- During the MX4SIO root sweep, a missing BDM entry means "not present yet".
+      -- Do not fall back to fileXioDopen while the mount thread may own the FS lock.
+      if mx4_mass_root_probe_budget > 0 then
+        return nil
+      end
+    end
+    return raw_get_mass_mount_driver(root)
+  end
```
[the remark this answers is not in the document]
Comment (NathanNeurotic) [reply etc/boot.lua]: Validated. The stale-snapshot risk was real, but the final fix no longer retains this wrapper/cache architecture. MX4SIO identity is now read fresh from `System.bdmList()` inside the authoritative `system.lua` discovery path and the boot-path translator. Outside those specific paths, `System.getMassMountDriver` is untouched. Regression coverage also verifies that MX4SIO discovery performs no bare-root filesystem or native mount-driver probes.

**Comment by NathanNeurotic** [comment]
@coderabbitai review

**Review by NathanNeurotic — COMMENTED** [review]
Superseded by #541. Hardware testing disproved this branch, and `dev` subsequently landed the broader storage-driver/ATA revert that must be the baseline. The replacement PR starts cleanly from that revert and retains only the independent MMCE, MX4SIO-discovery, staging, and handoff-test fixes.
